### PR TITLE
feat(extmark): support highlighting and querying multiline ranges

### LINKED ISSUE #17270: nvim_buf_get_extmarks doesn't return extmarks that contain than the requested span
### Neovim version (nvim -v)

NVIM v0.6.0

### Vim (not Nvim) behaves the same?

N/A

### Operating system/version

ArchLinux

### Terminal name/version

Running in neovide

### $TERM environment variable

xterm-256color

### Installation

Manjaro pacman

### How to reproduce the issue

Buffer contains:

> hello

Run

```
:let g:ns  = nvim_create_namespace('bug')
:call nvim_buf_set_extmark(0, g:ns, 0, 0, {"end_line": 0, "end_col": 4})
:echo nvim_buf_get_extmarks(0, g:ns, [0, 1], [0, 1], {})
```

### Expected behavior

Because the requested position is spanned by the `col` and `end_col` of the extmark, it should be returned: 

`[[1, 0, 0]]`

### Actual behavior

The extmark is not returned:

`[]`

### PR DESCRIPTION
**Pull request by bfredl**

### CODE REVIEW

**Comment by theHamsta** [comment]
This would be really useful to implement "locals" highlighting for tree-sitter. Just to look up the ranges a given point is currently contained in.

File: src/klib/kvec.h [hunk 1]
```diff
@@ -207,6 +207,16 @@ static inline void *_memcpy_free(void *const restrict dest, void *const restrict
   /* 2^x initial array size. */ \
   kvi_resize(v, (v).capacity << 1)
 
+/// fit at least "len" more items
+#define kvi_ensure_space(v, len) \
```
Comment (theHamsta) [inline src/klib/kvec.h:211]: the name suggests to me that it would ensure `len` in total, not additional `len` space

**Comment by bfredl** [comment]
This should be ready for testing. The user visible changes should be as described in news.txt 

>  Extmarks now fully supports multi-line ranges, and a single extmark can be
  used to highlight a range of arbitrary length. The |nvim_buf_set_extmark()|
  API function already allowed you to define such ranges, but highlight regions
  was not rendered consistently for a range that covers more than one line break.
  This has now been fixed. In addition, |nvim_buf_get_extmarks()| has gained
  a "overlap" option to return such ranges even if they started before the
  specified position.

**Comment by bfredl** [comment]
@lewis6991 I implemented all the `pending` tests for ranged signs. Note in one case I had to change the ordering of the the signs, does that look fine to you?

**Comment by lewis6991** [comment]
Yeah I think that's fine. Neither sign type defines a priority so the ordering shouldn't matter.

File: src/nvim/decoration.c [hunk 2]
```diff
@@ -412,6 +384,26 @@ void decor_redraw_signs(buf_T *buf, int row, int *num_signs, SignTextAttrs sattr
...
+    Decoration *decor = pair.start.decor_full;
+
+    if (!decor || !decor_has_sign(decor)) {
+      continue;
+    }
+
+    decor_to_sign(decor, num_signs, sattrs, num_id, line_id, cul_id);
+  }
+
   while (true) {
```
[the remark this answers is not in the document]
Comment (bfredl) [reply src/nvim/decoration.c:410]: > I guess the first loops covers "range marks" and the second loop handle "point marks"?

Basically yes. more specifically the first loop handles all range marks which overlaps the start position `(row, 0)`. the second loop handles ranges which start later in the line, as well as point marks.

Added a comment at `marktree_itr_get_overlap` as the "double loop" pattern is relevant to many call places

File: runtime/doc/news.txt [hunk 3]
```diff
@@ -219,6 +219,14 @@ The following changes to existing APIs or features add new behavior.
   "virtual_text" table, which gives users more control over how diagnostic
   virtual text is displayed.
 
+• Extmarks now fully supports multi-line ranges, and a single extmark can be
+  used to highlight a range of arbitrary length. The |nvim_buf_set_extmark()|
+  API function already allowed you to define such ranges, but highlight regions
+  was not rendered consistently for a range that covers more than one line break.
+  This has now been fixed. Signs defined as part of a multi-line extmark also
+  applies to every line in the range, not just the first.
+  In addition, |nvim_buf_get_extmarks()| has gained a "overlap" option to
```
Comment (zeertzjq) [inline runtime/doc/news.txt:228]:
```suggestion
  In addition, |nvim_buf_get_extmarks()| has gained an "overlap" option to
```

File: runtime/doc/news.txt [hunk 4]
```diff
@@ -219,6 +219,14 @@ The following changes to existing APIs or features add new behavior.
...
+  return such ranges even if they started before the specified position.
 ==============================================================================
```
Comment (zeertzjq) [inline runtime/doc/news.txt:230]:
```suggestion

==============================================================================
```

File: runtime/doc/news.txt [hunk 5]
```diff
@@ -221,6 +221,15 @@ The following changes to existing APIs or features add new behavior.
...
   virtual text is displayed.
 
+• Extmarks now fully supports multi-line ranges, and a single extmark can be
+  used to highlight a range of arbitrary length. The |nvim_buf_set_extmark()|
+  API function already allowed you to define such ranges, but highlight regions
+  was not rendered consistently for a range that covers more than one line break.
+  This has now been fixed. Signs defined as part of a multi-line extmark also
+  applies to every line in the range, not just the first.
+  In addition, |nvim_buf_get_extmarks()| has gained na "overlap" option to
+  return such ranges even if they started before the specified position.
```
Comment (clason) [inline runtime/doc/news.txt:231]:
```suggestion
• Extmarks now fully support multi-line ranges, and a single extmark can be
  used to highlight a range of arbitrary length. The |nvim_buf_set_extmark()|
  API function already allowed you to define such ranges, but highlight regions
  were not rendered consistently for a range that covers more than one line break.
  This has now been fixed. Signs defined as part of a multi-line extmark also
  apply to every line in the range, not just the first.
  In addition, |nvim_buf_get_extmarks()| has gained an "overlap" option to
  return such ranges even if they started before the specified position.
```

File: src/nvim/api/extmark.c [hunk 6]
```diff
@@ -451,6 +427,11 @@ Array nvim_buf_get_extmarks(Buffer buffer, Integer ns_id, Object start, Object e
 /// Using the optional arguments, it is possible to use this to highlight
 /// a range of text, and also to associate virtual text to the mark.
 ///
+/// If present, the position defined by `end_col` and `end_row` should be after
+/// the start position in order to the extmark to cover a range.
```
Comment (clason) [inline src/nvim/api/extmark.c:431]:
```suggestion
/// the start position in order for the extmark to cover a range.
```

**Comment by clason** [comment]
Almost three years in the making 🥳

**Pull request merged**